### DISCUSSION TITLE
[NET 8.0] BinaryFormatter disabled

### DISCUSSION THREAD

**Discussion by workgroupengineering** [AvaloniaUI/Avalonia#13470] [Q&A]
https://learn.microsoft.com/en-us/dotnet/core/compatibility/serialization/8.0/binaryformatter-disabled

**Comment by maxkatz6** [comment]
Currently, Avalonia only uses BinaryFormatter in DnD/clipboard code for WinForms compatibility. And only on Windows OS. So, it's not really a bit blocker for us.

We can expand it by introducing the same replacement as WinForms did - https://github.com/dotnet/winforms/pull/9088. Which replicates backward-compatible behavior of BinaryFormatter but without deprecated dependencies. At least it works for most of common types.

Unfortunately, it's a 4k lines replacement without any public packages that we can reuse.

**Reply by maxkatz6** [reply]
Also, to mention, I would want to avoid full winforms repo submodule. Subtree might work better here...

**Reply by maxkatz6** [reply]
Or we might as well drop winforms compatibility support at all, without any replacements. I believe it's hardly ever used, though i don't have any data to confirm.

**Reply by workgroupengineering** [reply]
Or separate repository with its own nuget package.

Can be uesed like this

```csharp
        public static AppBuilder BuildAvaloniaApp()
            => AppBuilder.Configure<App>()
                .UsePlatformDetect()
                .UseWinformInterop()
                .LogToTrace();
```

**Reply by maxkatz6** [reply]
Yeah, I think it's a possible solution. We can move this extension to the Avalonia.Win32.Interoperability project as well.

**Reply by maxkatz6** [reply]
"UseWinformInterop" naming is a bit confusing though, as it implies more than just a clipboard binary serialization.

**Reply by MrJul** [reply]
With 11.1 officially targeting .NET 8, I think it's time to tackle this issue.
My preference would be to remove the winforms clipboard compatibility support completely in 11.1 while it isn't too late to do so.

**Reply by maxkatz6** [reply]
11.1 is already in code freeze preparing for 11.1-beta early next week. Some bugfixes might be backported before 11.1 stable, and one more single requested feature that doesn't affect existing apps (Vulkan). Everything else will go to 11.1.1 (backports) or 11.2 (master + nightly).

**Comment by glowysourworm** [comment]
The funny part of this is their "security discussion":  https://devblogs.microsoft.com/dotnet/binaryformatter-removed-from-dotnet-9/

The risk of binary formatting is that it will take binary data and create some "malicious object" when deserialized!  Oh gosh! @_@  I never even knew binary data existed! I hope none of it is on my computer, or I'll get attacked by a "malicious object"! I hope this doesn't also happen when you use Stream, BitConverter, Array.CopyTo, shift operators, "unsafe" code, exceptions, native pointers, bytes, chars, strings,... ? No wait, not strings, right?  Those are the Alphabet! They're not "malicious"... 

FYI, I have a POC project for a Serializer replacement for .NET's BinaryFormatter, er... "Serializer", ISerializable, SerializableAttribute, etc.. etc...

There's many problems with this serializer - which I needed to work for structs - using depth-first complete serialization. Theirs, of course, does not tell you anything about how it deserializes the graph; and there are (sometimes) no exceptions thrown for when things get "weird". Mine happened to be a struct that returned "null" after trying to deserialize (probably constructor related) - with no exception, or reason as to why. The "bandaid" is their event hook for when "deserialization is complete". So, if you want to you can go through your graph and hook it up where the struct was finally finished (!!!) That's extremely poor design!

I was able to fix this, and found very many other topics for consideration:  Runtime.Type is a .NET boundary where these problems take shape. The Type class is not serializable. There are also issues trying to recurse. So, many different class AND object graphs (which are usually different - many object graphs may come from a single class graph) will NOT be serializable even though they compile. This indicates that the serialization problem is not complete.

I managed the following:  HashedType (serializable type wrapper); recursive class serialization; object graph difference checking (will support differences between the class and object graph - which would've arisen from changes to the code - which is normally not supported). 

So, for some situations, it would be a very good project to work on! It's functioning with only a couple weeks of work; and I'm currently using it for my AudioStation (new repository).

Anyway, good luck with Microsoft! For all the work you've put into Avalonia, I'd consider taking that one on - and be aware that you can own this problem. It was never done very well; and building tools for this one would be very useful. Why not have an object graph viewer where you can navigate the object tree, see what data is there, get / set data, save, serialize, create boiler plate, work with databases, etc... These tools would be a low hanging fruit; and already exist for most other serialization types.. just not the "malicious" ones (?)

(One more:  Protobuf) Anyone know the issue with protobuf having to use GUID's for "pointer references"? It means there will never be a way to have automatic object referencing! You must know how many / each "pointer" before hand.. right? You have to compile them in! The way to fix this is essentially what I did. As you serialize the data, you create references to each (complex) object. These must be sent along in the header for deserialization. I don't think they ever did this!

**Reply by stevemonaco** [reply]
`BinaryFormatter` was part of .NET 1.1 which was released back in 2003 when computing was a lot different. I'm not sure what your point is here as tech goes through churn, especially early designs. MS is stopping the pattern while still keeping a bridge that doesn't break 15-20+ year investments across the ecosystem. `BinaryFormatter` is a relatively easy footgun and it's bitten a lot of devs simply because of how unrestricted it is. https://learn.microsoft.com/en-us/dotnet/standard/serialization/binaryformatter-security-guide is probably a better discussion and there were a few across GitHub, too.
